test/retryOnEmpty.test.ts

```diff
@@ -0,0 +1,668 @@
+          await withTestSetup(async ({ engine, provider }) => {
+            const blockNumber = '0x0';
+            const request: JsonRpcRequest<string[]> = {
+              id: 1,
+              jsonrpc: '2.0',
+              method,
+              params: buildMockParamsWithBlockParamAt(
+                blockParamIndex,
+                blockNumber,
+              ),
+            };
+            const sendAsyncSpy = stubProviderRequests(provider, [
+              stubBlockNumberRequest(blockNumber),
+              stubRequestThatFailsThenFinallySucceeds({
+                request,
+                numberOfTimesToFail: 9,
+                successfulResponse: (req) => ({
+                  id: req.id,
+                  jsonrpc: '2.0',
+                  result: 'something',
+                }),
+              }),
+            ]);
+
+            const promiseForResponse = engine.handle(request);
+            await waitForRequestToBeRetried({
+              sendAsyncSpy,
+              request,
+              numberOfTimes: 10,
+            });
+
+            expect(await promiseForResponse).toStrictEqual({
+              id: 1,
+              jsonrpc: '2.0',
+              result: 'something',
+              error: undefined,
+            });
+          });
+        });
+
+        it('attempts the request up to 10 times, returning an error if it is still unsuccessful after that', async () => {
+          await withTestSetup(async ({ engine, provider }) => {
+            const blockNumber = '0x0';
+            const request: JsonRpcRequest<string[]> = {
+              id: 1,
+              jsonrpc: '2.0',
+              method,
+              params: buildMockParamsWithBlockParamAt(
+                blockParamIndex,
+                blockNumber,
+              ),
+            };
+            const sendAsyncSpy = stubProviderRequests(provider, [
+              stubBlockNumberRequest(blockNumber),
+              stubGenericRequest({
+                request,
+                response: (req) => {
+                  return {
+                    id: req.id,
+                    jsonrpc: '2.0',
+                    error: {
+                      code: -1,
+                      message: 'oops',
+                    },
+                  };
+                },
+                remainAfterUse: true,
+              }),
+            ]);
+
+            const promiseForResponse = engine.handle(request);
+            await waitForRequestToBeRetried({
+              sendAsyncSpy,
+              request,
+              numberOfTimes: 10,
+            });
+
+            expect(await promiseForResponse).toMatchObject({
+              error: expect.objectContaining({
+                message: 'RetryOnEmptyMiddleware - retries exhausted',
+              }),
+            });
+          });
+        });
+
+        it('falls through to the next middleware after making a request directly through the provider', async () => {
```

Comment:
This behavior is not intended. It's caused by another bug that was introduced around the same time - the same type of bug as #142 (and introduced in the same PR)

The final `next()` after a successful child request should not be there, because calling that results in the response from the child request being overwritten. It makes this whole middleware pointless.

Reply:
That's what I figured, but I wasn't sure 😵

Reply:
I've fixed this bug in this PR as well and updated the PR description to match.